### PR TITLE
Implement WikilinkProvider (DocLinkProvider + DefinitionProvider) to add the 'Go to Definition' command

### PR DESCRIPTION
**Pull request by JonasSprenger**
A try to replace `DocumentLinkProvider` by `DefinitionProvider`.
But for creating the file if the link is a `Placeholder`, the `DefinitionProvider` API don't seems right.

I don't think it's worth to change.

### CODE REVIEW

**Comment by JonasSprenger** [comment]
Original issuse #724

**Comment by riccardoferretti** [comment]
Thanks for looking into this @JonasSprenger!
At a high level, I like the feature, it would be good to have it in Foam if we can make it fit into our UX.

Would this provider replace the `DocumentLinkProvider` or does it work alongside it? is there anything we are "losing" by dropping the document link provider?

> But for creating the file if the link is a Placeholder, the DefinitionProvider API don't seems right.

Why is that, can you elaborate?

**Comment by JonasSprenger** [comment]
The call it's triggered on 'hover', so IMHO it's not adapted to create a new file if the link is a placeholder. Because it will necessarily happen inadvertently.

**Comment by JonasSprenger** [comment]
> Would this provider replace the `DocumentLinkProvider` or does it work alongside it? is there anything we are "losing" by dropping the document link provider?

Yes, keeping both will be confusing, so it's one or the other.

By dropping the document link provider we are "losing":
  - The ability to keep the cursor at the last position if the linked note is already open.
  - Keeping the creation of an new note from a placeholder explicit (see below).  

> > But for creating the file if the link is a Placeholder, the DefinitionProvider API don't seems right.
> 
> Why is that, can you elaborate?

The call to the `DefinitionProvider` is triggered on 'hover', so if the link is a placeholder an new file is triggered immediately.
It will necessarily happen inadvertently, I do not expect any notes to be created until a deliberated action or command (like `Cmd+Click`).
It will imply the same "surprising" behavior as with vscode-markdown-notes #715

I have two suggestion https://github.com/foambubble/foam/issues/543#issuecomment-881911226
  - There is the possibility to use the `Open Link` command if the link doesn't exist and `Go to Definition` if it exist. But I don't thinks it's worth the increase of complexity for the user (two different commands for almost the same functionality).
 - An alternative solution is to implement the `hover-feature` to have some of the benefits of `Go to Definition`.

**Comment by riccardoferretti** [comment]
If the two can coexist, I'd say let's have them coexist and use the right tool for the specific job, even if that means "splitting" functionality, no?
For code readability we could also create a `wikilink-providers.ts` that includes both the document link provider and the definition provider (as they are both related to wikilnks), do you think this would help achieve both goals (reduce confusion, full functionality)?

**Comment by JonasSprenger** [comment]
> If the two can coexist, I'd say let's have them coexist and use the right tool for the specific job, even if that means "splitting" functionality, no?

You're right, there must be a way to have them coexist, and it's worth to make it configurable :
  - Navigating only with Links `DocumentLinkProvider` (like actual behavior).
  - Navigating only with Definition `DefinitionProvider`.
  - Combining both by using `DefinitionProvider` if the note already exit and `DocumentLinkProvider` for placeholder (creating and open a new note).

**Comment by JonasSprenger** [comment]
> hi @JonasSprenger what is the status of this PR?

Hi @riccardoferretti,

I will try to add both Provider and make they coexist nicely with configuration.

**Comment by riccardoferretti** [comment]
if you rebase from master now, the issue you are having with the tests should be resolved

File: packages/foam-vscode/package.json [hunk 1]
```diff
@@ -235,10 +235,22 @@
...
+            "openLink",
+            "goToDefinitionButOpenLinkForCreation",
+            "goToDefinition",
+            "off"
+          ],
+          "enumDescriptions": [
+            "Use 'Open Link' command for navigation and creation",
+            "Use 'Go to Definition' command for navigation but use 'Open Link' for creation",
+            "Use 'Go to Definition' command for navigation and creation (Ctrl+hover over a placeholder will create and open the new note)",
+            "Disable navigation and creation"
```
Comment (JonasSprenger) [inline packages/foam-vscode/package.json:252]: What do you think about the formulation of the configuration options?

File: packages/foam-vscode/src/features/wikilink-provider.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      }
+
+      const newNoteUri = toVsCodeUri(
+        URI.createResourceUriFromPlaceholder(
+          vscode.workspace.workspaceFolders[0].uri,
+          uri
+        )
+      );
+      const newNoteTitle = uri.path.split('/').slice(-1)[0];
+      await createNoteForPlaceholderWikilink(newNoteTitle, newNoteUri);
```
Comment (JonasSprenger) [inline packages/foam-vscode/src/features/wikilink-provider.ts:94]: For creating a new note from a placeholder I used the same pattern as in `OPEN_COMMAND.execute` (`packages/foam-vscode/src/features/utility-commands.ts`), but only if a VsCode Workspace is open.

Is this the right way to go? @movermeyer  @riccardoferretti

File: packages/foam-vscode/src/features/wikilink-provider.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    const targetResource = this.workspace.get(uri);
+    const targetRange = findFirstLineOfContent(targetResource.source.text);
+
+    const result: vscode.LocationLink = {
+      originSelectionRange: toVsCodeRange(targetLink.range),
+      targetUri: toVsCodeUri(uri),
+      targetRange: null, // Disable the preview feature to avoid duplication with the over feature.
+      targetSelectionRange: toVsCodeRange(targetRange),
```
Comment (JonasSprenger) [inline packages/foam-vscode/src/features/wikilink-provider.ts:104]: When following a link with the `Go To Definition` command, this allows to briefly highlight the first line of the MD content (usually the title).

File: packages/foam-vscode/src/features/wikilink-provider.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    toVsCodeRange(link.range),
+    command
+  );
+  documentLink.tooltip = URI.isPlaceholder(target)
+    ? `Create note for '${target.path}'`
+    : `Go to ${URI.toFsPath(target)}`;
+  return documentLink;
+};
+
+export const findFirstLineOfContent = (rawContent: string): Range => {
```
Comment (JonasSprenger) [inline packages/foam-vscode/src/features/wikilink-provider.ts:149]: Is it OK to implement a custom parsing algorithm ? In the long term, it's maybe not the ideal place to stay.

I tried to use `targetResource.source.contentStart` but it's not sufficient, and it skips systematically two lines after the front matter if there is one (not sure if it's the intended behavior).

**Comment by JonasSprenger** [comment]
Hello @riccardoferretti,

I've finally come up with a reasonable solution (imho) to give the possibility to choose which command to use for opening links. With one possible option to combine both providers: `DefinitionProvider` to open existing notes and `DocumentLinkProvider` for creating a new note by `Ctrl+click` on a placeholder, the goal is to keep the creation of a new note (file) more explicit compared to `Ctrl+over` when using only `DefinitionProvider`.

Observation: When changing the configuration, the opens editors don't take the modifications in account, it is necessary to select another tab then re-select the original one for the changes to take effect (it's not a big deal in my opinion, but good to known when doing manual testing).

Apart from the missing tests, I have nothing to add or modify, so it's good for a review.

File: packages/foam-vscode/package.json (hunk 1 above)
Comment (riccardoferretti) [reply packages/foam-vscode/package.json:252]: What is the difference in behavior between `openLink` and `goToDefinitionButOpenLinkForCreation`?

As in, not technically, but what changes for the user?

File: packages/foam-vscode/src/features/wikilink-provider.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    token: vscode.CancellationToken
+  ): Promise<vscode.LocationLink[] | vscode.Definition> {
+    const navMode = this.getNavigationMode();
+    if (navMode === OPEN_LINK_NAV_MODE || navMode === NAVIGATION_DISABLED) {
+      return;
+    }
+
+    const resource = this.parser.parse(document.uri, document.getText());
+    const targetLink: ResourceLink | undefined = resource.links.find(link =>
+      Range.containsPosition(link.range, {
```
Comment (riccardoferretti) [inline packages/foam-vscode/src/features/wikilink-provider.ts:68]: can't you just use `Range.containsPosition(link.range, position)` here?

File: packages/foam-vscode/src/features/wikilink-provider.ts (hunk 2 above)
Comment (riccardoferretti) [reply packages/foam-vscode/src/features/wikilink-provider.ts:94]: My take here is that it's not this provider that should be concerned with the workspace status, but the command itself - as we'd want the behavior to be consistent across "callers"

what's the thinking behind the "exception"?

File: packages/foam-vscode/src/features/wikilink-provider.ts (hunk 3 above)
Comment (riccardoferretti) [reply packages/foam-vscode/src/features/wikilink-provider.ts:104]: Is this moving the cursor to that position? what if the note was already open? would it displace the former cursor position?

File: packages/foam-vscode/src/features/wikilink-provider.ts (hunk 4 above)
Comment (riccardoferretti) [reply packages/foam-vscode/src/features/wikilink-provider.ts:149]: Feels like a bug with `contentStart` (but I wouldn't fix it in this PR).

**Review by riccardoferretti — COMMENTED** [review]
Thanks for this, I left a couple of minor comments - the most important thing I want to fully understand is the effective change in behavior amongst the configuration options, so we can decide how flexible to be about them

File: packages/foam-vscode/src/features/wikilink-provider.ts (hunk 4 above)
Comment (JonasSprenger) [reply packages/foam-vscode/src/features/wikilink-provider.ts:149]: I agree, this PR it's not the place to fix it.

**Comment by riccardoferretti** [comment]
Hi @JonasSprenger, based on this PR I have created https://github.com/foambubble/foam/pull/821

Let me know if you have any comment there, and thanks for spear-heading the effort (it was easy for me to implement that based on the work you did here!)

Closing this PR as it is now replaced by #821

**Comment by JonasSprenger** [comment]
Hi @riccardoferretti,

Sorry for the long delays of my contributions and my responses. I was struggling and I didn't achieve to make a unit test for the creation of a new note within the DefinitionProvider. But most of all I struggle with procrastination... :unamused:

I'm glad my contribution was still useful.

Concerning your implementation, I like the pragmatic approach to not bother with configurability, and +1 to provide References of the Backlinks.

**Pull request closed without merging**